### PR TITLE
Provide parse options to the language service more quickly

### PR DESCRIPTION
**Pull request by jasonmalinowski**
When we load a project in Visual Studio, the project system first does an evaluation pass of the project, and hands the resulting list of <Compile> items to the language service. It then does an execution pass executing CoreCompile passing SkipCompilerExecution=true and ProvideCommandLineArgs=true, that resulting command line string is where we get our compiler switches. The execution pass is much slower than the evaluation pass, so there's a window of time where we have a list of files, but not any options yet.

Because there's a gap, that means there's a time where we are parsing source files with the default parse options. We'll then have to reparse them a second time which isn't great. It also means any cache lookups we do won't have the right options either, so the cache lookups might miss.

To help this, we'll have a property for the evaluation pass which is an "approximation" of the options that would come out of CoreCompile, but only the ones that matter for parsing. It's acceptable for this to be imperfect: once the execution pass is complete we'll use those options instead, so any behaviors here that don't match the real command line generation will only be temporary, and probably won't be any worse than having no options at all.

This is the targets side of https://github.com/dotnet/roslyn/issues/60014; the project system changes were already made and have been merged.

### CODE REVIEW

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets [hunk 1]
```diff
@@ -161,4 +161,26 @@
...
+
+       To help this, we'll have a property for the evaluation pass which is an "approximation" of the
+       options that would come out of CoreCompile, but only the ones that matter for parsing. It's acceptable
+       for this to be imperfect: once the execution pass is complete we'll use those options instead,
+       so any behaviors here that don't match the real command line generation will only be temporary, and
+       probably won't be any worse than having no options at all. -->
+       
+  <PropertyGroup>
+    <CommandLineArgsForDesignTimeEvaluation>-langversion:$(LangVersion)</CommandLineArgsForDesignTimeEvaluation>
+    <CommandLineArgsForDesignTimeEvaluation Condition="'$(DefineConstants)' != ''">$(CommandLineArgsForDesignTimeEvaluation) -define:$(DefineConstants)</CommandLineArgsForDesignTimeEvaluation>
```
Comment (jasonmalinowski) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:184]: One issue that was identified here was the SDK computes some of its constants during execution, which means this optimization doesn't always work as ideally as we'd like. There's other problems with that and we'll investigate further in what it'd take to do. We have some recently added optimizations around avoiding reparsing if there's only a preprocessor directive change and the file doesn't have any preprocessor directives; having us at least get the langversion over means we have a better chance of that optimization being able to kick in.

tl;dr; this doesn't work in all cases, but that's better than working on no cases.

Comment (RikkiGibson) [reply src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:184]: just curious, is there any telemetry for when "misses" occur here? something we could look at occasionally and say--oops, we added this option which affects parsing, we can see that we need to update this target.

**Review by RikkiGibson — APPROVED** [review]
changes under Compilers LGTM

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-ide, @dotnet/roslyn-compiler this needs one more review from each team.

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets (hunk 1 above)
Comment (jasonmalinowski) [reply src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:184]: We do have telemetry that we added when we initially thought of doing this: https://github.com/dotnet/roslyn/blob/c756bbabff2344cc1813c968245d8477a9ae3673/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs#L280. My concern isn't the lack of the telemetry, but how we'll know to actually look at it if/when the time comes. I admit I don't have any good ideas there.

**Review by genlu — APPROVED** [review]

File: src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets [hunk 2]
```diff
@@ -134,4 +134,26 @@
...
+
+       To help this, we'll have a property for the evaluation pass which is an "approximation" of the
+       options that would come out of CoreCompile, but only the ones that matter for parsing. It's acceptable
+       for this to be imperfect: once the execution pass is complete we'll use those options instead,
+       so any behaviors here that don't match the real command line generation will only be temporary, and
+       probably won't be any worse than having no options at all. -->
+       
+  <PropertyGroup>
+    <CommandLineArgsForDesignTimeEvaluation>-langversion:$(LangVersion)</CommandLineArgsForDesignTimeEvaluation>
+    <CommandLineArgsForDesignTimeEvaluation Condition="'$(FinalDefineConstants)' != ''">$(CommandLineArgsForDesignTimeEvaluation) -define:$(FinalDefineConstants)</CommandLineArgsForDesignTimeEvaluation>
```
Comment (333fred) [inline src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets:157]: Why are the property names different between C# and VB?

Comment (jasonmalinowski) [reply src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets:157]: Seems to be a history thing. If you look at the Csc task vs. Vbc task, there's a different input that's given to them. And I see this too:

https://github.com/dotnet/sdk/blob/2693567468cbaf97c85c57e277e2a0609b317e46/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.VisualBasic.targets#L55

Comment (333fred) [reply src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets:157]: Sigh

**Review by 333fred — APPROVED** [review]
